### ISSUE TITLE
Run annotators during export/snapshot

### ISSUE THREAD

**Issue by minad** [oantolin/embark#203]
This way you always get annotations even if the minibuffer is closed. And I suspect it would also be less problematic regarding the annotation function requirements (default-directory, target, ...).

**Comment by oantolin** [comment]
Export has no use for annotations, there is nowhere to put them. So let's forget about export.

I take it you are proposing that Embark collect snapshot buffers compute the annotations once when they are created and then just store them? I guess they could do that. But that strategy doesn't really work for live collect buffers or collect completions buffers. Since I needed to implement recomputing annotations upon revert for those buffer anyway, I just use the same code for snapshot buffers.

**Comment by minad** [comment]
> I take it you are proposing that Embark collect snapshot buffers compute the annotations once when they are created and then just store them? I guess they could do that. But that strategy doesn't really work for live collect buffers or collect completions buffers. Since I needed to implement recomputing annotations upon revert for those buffer anyway, I just use the same code for snapshot buffers.

Yes. Why does it not work for live collect? When collecting for live collect you also run the annotators directly. The only downside I see is that you don't want to collections the annotations when using the grid view.

My proposal would be to store the annotations in `embark-collect-candidates`. Each candidate can either be a string or a list with two or three elements (like the affixation function specifies).

**Comment by oantolin** [comment]
I'm sorry, I think i misunderstood you. The annotations are currently computed when `embark-collect-candidates` whenever is updated. I thought you were proposing to change when they were computed, but you are only proposing to change where they are stored?

**Comment by minad** [comment]
>  The annotations are currently computed whenever embark-collect-candidates is updated (and you are in list view). 

Yes, I know. This means it does not work when the minibuffer is closed.

> I thought you were proposing to change when they were computed, but you are only proposing to change where they are stored (and to also compute them in grid view)?

I propose to compute annotations directly when collecting the candidates. Then the minibuffer is still open and we don't lose them. Second, I propose to store the annotations together with the candidates in the `embark-collect-candidates` list in the three variants I mentioned, `cand`, `(cand suffix)`, `(cand prefix suffix)`. The cand-prefix-suffix list elements are returned as such by the affixation function.

**Comment by oantolin** [comment]
Ah, sorry. This whole conversation I was laboring under the mistaken impression that Embark stored the annotation-function while the minibuffer was still open. I see your point now. I'm sorry for confusion. I think I did use to call revert before the minibuffer was closed, which is why I thought I still did.

**Comment by oantolin** [comment]
Since almost all my annotations come from Marginalia and Embark uses those as a fallback, I didn't realize I wasn't capturing the annotation function (anymore?, I think I used to).

**Comment by oantolin** [comment]
I took a first step: capturing the annotator when you call embark collect. f52f9a34893a0dda493e22326aa859545a56bfe8

I guess this way I do have the annotators but they might run in the wrong buffer.

**Comment by oantolin** [comment]
I didn't test the code, only read it very superficially, but I believe you run the affixator upon collection and then never again? That breaks marginalia-cycle in embark collect buffers (which I don't actually use, but some people might). Also, I didn't see the part where you strip the annotations in embark-export, won't dired and ibuffer choke on them?

**Comment by oantolin** [comment]
And, yes I'd rather not calculate the affixations or annotations at all in grid view.

**Comment by oantolin** [comment]
Annotators and affixators now are run at snapshot time, see 21d4ca14e343657e44203a73edcac9edb39f0679.
